### PR TITLE
Dao 8 servlet 2

### PR DESCRIPTION
**Pull request by Ketch92**

### CODE REVIEW

File: src/main/webapp/WEB-INF/views/car/add_driver_to_car.jsp [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+<html>
+<head>
+    <title>Add driver</title>
+</head>
+<body>
+<h1>Please, fill in the car info.</h1>
+
+<h4 style="color: red">${errorMessage}</h4>
+
+<form method="post" action=${pageContext.request.contextPath}"/car/add/driver">
```
Comment (gigbat) [inline src/main/webapp/WEB-INF/views/car/add_driver_to_car.jsp]: It's up to you, but you could do such
```suggestion
<form method="post" action="${pageContext.request.contextPath}/car/add/driver">
```

File: src/main/webapp/WEB-INF/views/car/drivers_of_a_car.jsp [hunk 2]
```diff
@@ -0,0 +1,35 @@
+
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+
```
Comment (gigbat) [inline src/main/webapp/WEB-INF/views/car/drivers_of_a_car.jsp]: as i know here shouldn't add empty line

File: src/main/webapp/WEB-INF/views/index.jsp [hunk 3]
```diff
@@ -1,9 +1,32 @@
...
 <html>
 <head>
-    <title>Title</title>
+    <title>Taxi service</title>
 </head>
 <body>
-<h1>Hello World</h1>
+<h1>Taxi service</h1>
+
+<p>
```
Comment (gigbat) [inline src/main/webapp/WEB-INF/views/index.jsp:9]: tag <p> need to describe block with text, but if it just link without text you need not to add tag <p>

File: src/main/webapp/WEB-INF/views/manufacturer/add.jsp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+<form method="post" action=${pageContext.request.contextPath}"/manufacturer/add">
+    Manufacturer's name <input type="text" name="name">
+    Manufacturer's country <input type="text" name="country">
+    <button type="submit">Send</button>
+</form>
+<p>
+    <a href=${pageContext.request.contextPath}"/">Return to main page</a>
+</p>
+</body>
+</html>
```
Comment (gigbat) [inline src/main/webapp/WEB-INF/views/manufacturer/add.jsp]: add empty line

File: src/main/webapp/web.xml [hunk 5]
```diff
@@ -3,13 +3,84 @@
...
+    <servlet>
+        <servlet-name>addDriverToCar</servlet-name>
+        <servlet-class>core.controller.car.AddDriverToCarController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>addDriverToCar</servlet-name>
+        <url-pattern>/car/add/driver</url-pattern>
     </servlet-mapping>
 </web-app>
```
Comment (gigbat) [inline src/main/webapp/web.xml:127]: add empty line

**Review by gigbat — CHANGES_REQUESTED** [review]
Perfect job

File: src/main/webapp/web.xml (hunk 5 above)
Comment (Ketch92) [reply src/main/webapp/web.xml:127]: Did

File: src/main/webapp/WEB-INF/views/manufacturer/add.jsp (hunk 4 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/manufacturer/add.jsp]: added

File: src/main/webapp/WEB-INF/views/car/drivers_of_a_car.jsp (hunk 2 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/car/drivers_of_a_car.jsp]: removed

File: src/main/webapp/WEB-INF/views/car/add_driver_to_car.jsp (hunk 1 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/car/add_driver_to_car.jsp]: fixed

File: src/main/webapp/WEB-INF/views/index.jsp (hunk 3 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/index.jsp:9]: fixed

**Review by gigbat — APPROVED** [review]

File: src/main/java/core/controller/car/AddCarController.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/car/add.jsp").forward(req, resp);
+    }
+    
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String model = req.getParameter("model");
+        String manufacturerName = req.getParameter("manufacturerName");
+        String manufacturerCountry = req.getParameter("manufacturerCountry");
```
Comment (KseniiaMakarova) [inline src/main/java/core/controller/car/AddCarController.java]: we can just get the id of the manufacturer that is already present in the DB

File: src/main/java/core/controller/car/AddDriverToCarController.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        try {
+            Driver driver = driverService.get(Long.valueOf(driverId));
+            Car car = carService.get(Long.valueOf(carId));
+            carService.addDriverToCar(driver, car);
+            resp.sendRedirect(req.getContextPath() + "/");
+        } catch (NoSuchElementException e) {
+            errorInput(req, resp, "Such car or drivers isn't registered");
+        }
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/controller/car/AddDriverToCarController.java]: redundant empty line

File: src/main/java/core/controller/car/AddDriverToCarController.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        req.getRequestDispatcher("/WEB-INF/views/car/add_driver_to_car.jsp").forward(req, resp);
+    }
+    
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String carId = req.getParameter("carId");
+        String driverId = req.getParameter("driverId");
+        if (carId.isEmpty() || driverId.isEmpty()) {
+            errorInput(req, resp, "Empty input provided");
```
Comment (KseniiaMakarova) [inline src/main/java/core/controller/car/AddDriverToCarController.java]:
```suggestion
            errorInput(req, resp, "Empty input provided");
            return;
```
we should not be able to go further in this method

File: src/main/java/core/controller/car/AddDriverToCarController.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        try {
+            Driver driver = driverService.get(Long.valueOf(driverId));
+            Car car = carService.get(Long.valueOf(carId));
+            carService.addDriverToCar(driver, car);
+            resp.sendRedirect(req.getContextPath() + "/");
+        } catch (NoSuchElementException e) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/controller/car/AddDriverToCarController.java]: this is not a good practice to catch RuntimeExceptions
let's throw some custom checked Exception on the service layer instead - or we can just remove this check and assume that ids should be correct (this is even a better option)

File: src/main/java/core/utils/ConnectionUtils.java [hunk 10]
```diff
@@ -6,6 +6,15 @@
 import java.util.Properties;
 
 public class ConnectionUtils {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/utils/ConnectionUtils.java]: redundant empty line

File: src/main/webapp/web.xml [hunk 11]
```diff
@@ -3,13 +3,125 @@
...
     </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addDriver</servlet-name>
+        <servlet-class>core.controller.driver.AddDriverController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>addDriver</servlet-name>
+        <url-pattern>/driver/add</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/drivers/add</url-pattern>
```
it is recommended to use plural names in URLs

File: src/main/webapp/web.xml [hunk 12]
```diff
@@ -3,13 +3,125 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>getAllDrivers</servlet-name>
+        <servlet-class>core.controller.driver.GetAllDriversController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>getAllDrivers</servlet-name>
+        <url-pattern>/driver/all</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/drivers/</url-pattern>
```

File: src/main/webapp/web.xml [hunk 13]
```diff
@@ -3,13 +3,125 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>getAllManufacturers</servlet-name>
+        <servlet-class>core.controller.manufacturer.GetAllManufacturersController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>getAllManufacturers</servlet-name>
+        <url-pattern>/manufacturer/all</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]: same here

File: src/main/webapp/web.xml [hunk 14]
```diff
@@ -3,13 +3,125 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>getAllCars</servlet-name>
+        <servlet-class>core.controller.car.GetAllCarsController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>getAllCars</servlet-name>
+        <url-pattern>/car/all</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]: and here

File: src/main/webapp/web.xml [hunk 15]
```diff
@@ -3,13 +3,125 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addDriverToCar</servlet-name>
+        <servlet-class>core.controller.car.AddDriverToCarController</servlet-class>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>addDriverToCar</servlet-name>
+        <url-pattern>/car/add/driver</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/cars/drivers/add</url-pattern>
```

File: src/main/webapp/WEB-INF/views/car/add.jsp [hunk 16]
```diff
@@ -0,0 +1,19 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Add driver</title>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/add.jsp]:
```suggestion
    <title>Add car</title>
```

File: src/main/webapp/WEB-INF/views/car/drivers_of_a_car.jsp [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/drivers_of_a_car.jsp]: redundant empty line

File: src/main/webapp/WEB-INF/views/car/drivers_of_car.jsp [hunk 18]
```diff
@@ -0,0 +1,34 @@
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>All drivers</title>
+</head>
+<body>
+<h1>All drivers</h1>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/drivers_of_car.jsp]: what is the difference between this JSP and `car/drivers_of_a_car`?

File: src/main/webapp/WEB-INF/views/car/get_all.jsp [hunk 19]
```diff
@@ -0,0 +1,49 @@
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/get_all.jsp]: remove the empty line here in all the JSPs

File: src/main/webapp/WEB-INF/views/manufacturer/add.jsp [hunk 20]
```diff
@@ -0,0 +1,18 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Add driver</title>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/manufacturer/add.jsp]: fix title please

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/controller/car/AddCarController.java (hunk 6 above)
Comment (Ketch92) [reply src/main/java/core/controller/car/AddCarController.java]: Sure, changed

File: src/main/java/core/controller/car/AddDriverToCarController.java (hunk 7 above)
Comment (Ketch92) [reply src/main/java/core/controller/car/AddDriverToCarController.java]: True, thx

File: src/main/java/core/controller/car/AddDriverToCarController.java (hunk 8 above)
Comment (Ketch92) [reply src/main/java/core/controller/car/AddDriverToCarController.java]: I thought of that redirect in errorInput() method breaks the further execution of the method. Thanks, good to know

File: src/main/java/core/controller/car/AddDriverToCarController.java (hunk 9 above)
Comment (Ketch92) [reply src/main/java/core/controller/car/AddDriverToCarController.java]: Ok, then I refactored the method to the second option, as the first one requires changes in the service layer and all the interfaces and that isn't SOLID

File: src/main/java/core/utils/ConnectionUtils.java (hunk 10 above)
Comment (Ketch92) [reply src/main/java/core/utils/ConnectionUtils.java]: removed

File: src/main/webapp/web.xml (hunk 11 above)
Comment (Ketch92) [reply src/main/webapp/web.xml]: A bit of refactoring and we are there

File: src/main/webapp/web.xml (hunk 12 above)
Comment (Ketch92) [reply src/main/webapp/web.xml]: changed

File: src/main/webapp/web.xml (hunk 13 above)
Comment (Ketch92) [reply src/main/webapp/web.xml]: changed

File: src/main/webapp/web.xml (hunk 14 above)
Comment (Ketch92) [reply src/main/webapp/web.xml]: also changed

File: src/main/webapp/web.xml (hunk 15 above)
Comment (Ketch92) [reply src/main/webapp/web.xml]: Ok, changed

File: src/main/webapp/WEB-INF/views/car/add.jsp (hunk 16 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/car/add.jsp]: Copy-paste in power

File: src/main/webapp/WEB-INF/views/car/drivers_of_car.jsp (hunk 18 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/car/drivers_of_car.jsp]: there are different title and the second one returns user to all cars list while first returns to index.jsp

File: src/main/webapp/WEB-INF/views/car/get_all.jsp (hunk 19 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/car/get_all.jsp]: Done

File: src/main/webapp/WEB-INF/views/manufacturer/add.jsp (hunk 20 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/manufacturer/add.jsp]: Did, thanks and sorry

File: src/main/java/core/controller/car/AddDriverToCarController.java [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+            throws ServletException, IOException {
+        req.getRequestDispatcher("/WEB-INF/views/car/add_driver_to_car.jsp").forward(req, resp);
+    }
+    
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String carId = req.getParameter("carId");
+        String driverId = req.getParameter("driverId");
+        if (carId.isEmpty() || driverId.isEmpty()) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/controller/car/AddDriverToCarController.java]: we can also remove this and do the validation on the JSP

File: src/main/webapp/WEB-INF/views/car/add_driver_to_car.jsp [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+    <title>Add driver</title>
+</head>
+<body>
+<h1>Please, fill in the car info.</h1>
+
+<h4 style="color: red">${errorMessage}</h4>
+
+<form method="post" action="${pageContext.request.contextPath}/cars/drivers/add">
+    Car's id <input required type="number" name="carId">
+    Driver id <input required type="number" name="driverId">
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/add_driver_to_car.jsp:11]: here - if the input is required, can it be empty?

File: src/main/java/core/controller/driver/DeleteDriverController.java [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+            = Injector.getInstance("core");
+    private final DriverService driverService
+            = (DriverService) injector.getInstance(DriverService.class);
+    
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String id = req.getParameter("id");
+        driverService.delete(Long.valueOf(id));
+        resp.sendRedirect(req.getContextPath() + "/drivers/all");
```
Comment (KseniiaMakarova) [inline src/main/java/core/controller/driver/DeleteDriverController.java]:
```suggestion
        resp.sendRedirect(req.getContextPath() + "/drivers/");
```

File: src/main/webapp/WEB-INF/views/car/add.jsp [hunk 24]
```diff
@@ -0,0 +1,20 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Add car</title>
+</head>
+<body>
+<h1>Please, fill in the car info.</h1>
+
+<h4 style="color: red">${errorMessage}</h4>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/add.jsp]: probably we don't need this, same for other JSPs

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
added only a few remarks, good job!

File: src/main/java/core/controller/car/AddDriverToCarController.java (hunk 21 above)
Comment (Ketch92) [reply src/main/java/core/controller/car/AddDriverToCarController.java]: removed this and added require on jsps

File: src/main/webapp/WEB-INF/views/car/add_driver_to_car.jsp (hunk 22 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/car/add_driver_to_car.jsp:11]: solved

File: src/main/java/core/controller/driver/DeleteDriverController.java (hunk 23 above)
Comment (Ketch92) [reply src/main/java/core/controller/driver/DeleteDriverController.java]: Hmmm... i have been chacking all the url few times, still missed this one. Thanks

File: src/main/webapp/WEB-INF/views/car/add.jsp (hunk 24 above)
Comment (Ketch92) [reply src/main/webapp/WEB-INF/views/car/add.jsp]: Now we don't

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**